doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+    ResolveAccess.kRead, context)
+
+isLocal = LocatableContentTrait(resolved_asset).getIsLocal()  # May be None
+```
+Thankfully, we have already accounted for this. Our traitViews
+understand missing properties, and return `None`/`Optional`, so the host
+will know that there is no information on this new property from the
+current version of the manager.
+
+# Versioning Specifications
+Traits are not all we have to consider, lets take a moment to think
+about specifications.
+
+```yml
+# For example, Workfile, A specification with the Entity, Work, and LocatableContent traits.
+Workfile:
+  description: >
+    The entity is the product of some manual task or process,
+    that often defines how to produce or derive other data.
+
+
+    Common examples are the documents opened, worked on and saved
+    by artists using DCC tools to create models/images/etc.
+  usage:
+    - entity
+  traitSet:
+    - namespace: usage
+      name: Entity
+    - namespace: application
+      name: Work
+    - namespace: content
+      name: LocatableContent
+```
+
+Conceptually, specifications are "things" that exhibit a set of traits.
+Current, specifications are purely unordered sets of `traitId`'s, raising similar problems as how to encode the version that individual traits have.
+
+Specifications are used to categorize entities when performing api operations. For example, they may be used to filter an asset browser to specific kinds of entity (Image, Video, TextDoc, etc). They are also used in publishing, where both host and manager have to agree on the "kind of thing" being published.
+
+Initially, one would want to version these things with the same
+mechanism to traits, with the version being updated when the `traitSet`
+of the specification is modified. However, what happens if a trait in
+the specification is merely versioned up, should this imply a
+specification schema version increment?
+
+Tentatively, no, or at least, it would be better if it didn't. The
+method of versioning plays into this decision quite strongly:
+
+### Versioning entire schema
+- Has the advantage that version bumps are always together, so if a
+  trait updates, the specification implicitly updates with it.
+- Much simpler to think about, the specification version in-play is
+  always the one the manager reports.
+- Don't have to find a way to pack specification version into `traitSet`
+- Makes discoverability extra difficult for managers that are trying to
+  support ranges of compatible versions with fewer branches.
+
+### Versioning specs individually
+- Raises the question of whether each trait in a specification needs to
+  be pinned to a specific version of a trait.
+- Have to pack specification version into the type, or ask the manager
+  to somehow deduce specification version
+
+The complexity of thinking about disconnected versioning when you take
+specifications into account does seem to bias towards versioning the
+entire schema as a unit.
```

Comment:
As long as the spec and trait versioning is handled separately, I would advocate the opposite.  The per spec version does not have to be effected by trait versions.

Reply:
The ideal solution would indeed have Specifications continue to work regardless of changes in the traits making up their trait set.

There are technical reasons why that's difficult
* Specifications currently have no identity of their own in the API - i.e. when you make an API query you never provide the name of the Specification you're using, just its constituent traits. So there's currently no mechanism for the manager to know what Specification was used by the host at all, let alone the version of it. Adding such information would be a fairly big API rethink, but it is something we've discussed.
* The Specification classes generated from the YAML schema (using `traitgen`) currently have a hard dependency on the classes generated for their Traits, meaning mismatched Trait and Specification versions could lead to false-negative queries, or worse. This dependency is something we're considering removing as a consequence.